packages/pyright-internal/src/baseline.ts

```diff
@@ -0,0 +1,94 @@
+        }[];
+    };
+}
+
+export const baselineFilePath = (rootDir: Uri) => rootDir.combinePaths('.basedpyright/baseline.json');
+
+export const diagnosticsToBaseline = (rootDir: Uri, filesWithDiagnostics: FileDiagnostics[]): BaselineFile => {
+    const baselineData: BaselineFile = {
+        files: {},
+    };
+    for (const fileWithDiagnostics of filesWithDiagnostics) {
+        const filePath = rootDir.getRelativePath(fileWithDiagnostics.fileUri)!.toString();
+        if (!fileWithDiagnostics.diagnostics.length) {
+            continue;
+        }
+        if (!(filePath in baselineData.files)) {
+            baselineData.files[filePath] = [];
+        }
+        baselineData.files[filePath].push(
+            ...fileWithDiagnostics.diagnostics.map((diagnostic) => ({
+                code: diagnostic.getRule() as DiagnosticRule | undefined,
+                message: diagnostic.message,
+                range: diagnostic.range,
+            }))
+        );
+    }
+    return baselineData;
+};
+
+export const writeBaselineFile = (rootDir: Uri, baselineData: BaselineFile) => {
+    const baselineFile = baselineFilePath(rootDir);
+    mkdirSync(baselineFile.getDirectory().getPath(), { recursive: true });
+    writeFileSync(baselineFile.getPath(), JSON.stringify(baselineData, undefined, 4));
+};
+
+export const writeDiagnosticsToBaselineFile = async (rootDir: Uri, filesWithDiagnostics: FileDiagnostics[]) => {
+    const baselineData = diagnosticsToBaseline(rootDir, filesWithDiagnostics);
+    writeBaselineFile(rootDir, baselineData);
+};
+
+export const getBaselinedErrors = (rootDir: Uri): BaselineFile => {
+    const path = baselineFilePath(rootDir).getPath();
+    let baselineFileContents;
+    try {
+        baselineFileContents = readFileSync(path, 'utf8');
+    } catch (e) {
+        return { files: {} };
+    }
+    return JSON.parse(baselineFileContents);
+};
+
+export const filterOutBaselinedDiagnostics = (
+    rootDir: Uri,
+    filesWithDiagnostics: readonly FileDiagnostics[]
+): FileDiagnostics[] => {
+    const baselineFile = getBaselinedErrors(rootDir);
+    return filesWithDiagnostics.map((fileWithDiagnostics) => {
+        const baselinedErrorsForFile =
+            baselineFile.files[rootDir.getRelativePath(fileWithDiagnostics.fileUri)!.toString()];
+        if (!baselinedErrorsForFile) {
+            return fileWithDiagnostics;
+        }
+        return {
+            ...fileWithDiagnostics,
+            diagnostics: fileWithDiagnostics.diagnostics.filter((diagnostic) => {
+                const matchedIndex = baselinedErrorsForFile.findIndex(
+                    (baselinedError) =>
+                        baselinedError.message === diagnostic.message &&
+                        baselinedError.code === diagnostic.getRule() &&
+                        baselinedError.range.start.character === diagnostic.range.start.character &&
+                        baselinedError.range.end.character === diagnostic.range.end.character
```

Comment:
this matching strategy probably sucks. need to "battle test" this against a real codebase

Reply:
using the error message to match isn't a good idea because the wording of error messages changes often and we don't want the baselined errors to not match if a user is using basedpyright in a different language to what the baseline file was written with

Reply:
remove the line number if we arent matching with it, as it can cause needlessly large diffs

Reply:
https://github.com/DetachHead/basedpyright/pull/608/commits/1f1b88a9a15f9ace614dd8b1e20f506ab54fff42

Reply:
prob still need to inspect line numbers, eg. when imports change it's common for baselined deprecation errors to incorrectly come back

Reply:
actually in most cases if code has changed on the same line as an error you weant it to be unbaselined, i dont think theres a way arouind this without special casing imports, which i dont want to do